### ISSUE TITLE
Modernisation of 2MGFX

### ISSUE THREAD

**Issue by JoaoBaptMG** [MonoGame/MonoGame#6968]
Researching the state of the art of cross-compilation of HLSL shaders to other languages, I saw that the [Microsoft DirectX Shader Compiler](https://github.com/Microsoft/DirectXShaderCompiler) is now open source and can be run from non-Windows platforms, and that, in addition to DXIL, it can also emit SPIR-V, which can be cross compiled to GLSL, ESSL and possibly the console-specific shader languages (or rather kept as SPIR-V for those, since every platform that supports Vulkan supports SPIR-V). This sparked in me a possibility of modernisation of the shader compiler tool to be able to target higher shader models (in a preparation of MonoGame to modern shader and GPU features), as well as finally turning it cross-platform.

I do not have a clear/fixed roadmap for it, but I figure out that the required steps for it would be:

1. Figure out how to reference fxc in a cross-platform way, since it’s a native tool, not a .NET tool.
2. Figure out how to use [SPIRV-Cross](https://github.com/KhronosGroup/SPIRV-Cross) to emit GLSL for the OpenGL platforms.
3. Rewrite the 2MGFX tool to use both, emitting DXIL for the DX platforms, GLSL/ESSL for the OpenGL/ES platforms, MSL for a possible iOS renderer, and SPIR-V for console/Vulkan-enabled platforms?

**Comment by Jjagg** [comment]
Yes, the plan with regards to shader compilation has been since the DX Shader Compiler has been made public to use it in combination with SPIRV-Cross to reach all targets, however no meaningful work has been done in this area (AFAIK). For the Effects Framework stuff we want to use [HlslTools](https://github.com/tgjones/HlslTools) to parse and strip the Effects metadata from the fx files before passing them to the compiler pipeline.

> Figure out how to reference fxc in a cross-platform way, since it’s a native tool, not a .NET tool.

Why? We don't need fxc with the suggested pipeline. If you do need to run it, it works on Wine.

> so it might be useful to add a way to attach shaders to the render state using the new way as well

You mean attaching individual shaders rather than a full pipeline? A lower level shader API has come up several times over the years and it's definitely something I'd like MG to have. See e.g. #3471.

**Comment by JoaoBaptMG** [comment]
@Jjagg I'll pick up to work on the 2MGFX thing first, *then* I'll think a way to make MonoGame profit from the new API features. I was in look mainly for read/write objects, structured buffers and UAVs, but having subroutines (whatever they're called in DX) and other things will be nice.

**Comment by JoaoBaptMG** [comment]
The idea about attaching individual stages can be used in complement to the Effect API, making, for example Effects usable with SM 4.0 and 5.0, and having they attach geometry/tesselation shaders, as well as specifying they want to stream-out, instead of letting them keep rendering, and other things more.

**Comment by JoaoBaptMG** [comment]
> Why? We don't need fxc with the suggested pipeline. If you do need to run it, it works on Wine.
Oops, my bad, I wanted to say dxc.

**Comment by KakCAT** [comment]
>     2\. Figure out how to use [SPIRV-Cross](https://github.com/KhronosGroup/SPIRV-Cross) to emit GLSL for the OpenGL platforms.

Although it's not very active lately, check this project: https://github.com/microsoft/ShaderConductor

**Comment by JoaoBaptMG** [comment]
> 
> 
>
> > 2\. Figure out how to use [SPIRV-Cross](https://github.com/KhronosGroup/SPIRV-Cross) to emit GLSL for the OpenGL platforms.
>
> 
> Although it's not very active lately, check this project: https://github.com/microsoft/ShaderConductor

Yes, I picked up the idea of interfacing DXC with SPIRV-Cross exactly from there. I found a small problem in the DXC though:
> dxc.exe, a command-line tool that can compile HLSL programs *for shader model 6.0 or higher*
Maybe I have read wrong, but DXC cannot be used to build shaders on older shader models.

**Comment by Jjagg** [comment]
The only syntax that's no longer supported is some of the SM1.0 syntax (e.g. the tex2d intrinsic). Other than that it's backwards compatible.
Moving forward we intend to remove or deprecate the old syntax by parsing the old syntax with HlslTools and throwing errors or raising warnings and replacing syntax where possible (though I would personally prefer just throwing errors).

**Comment by Jjagg** [comment]
Oh, and SPIRV-CROSS can compile to older HLSL Shader Models.

**Comment by Jjagg** [comment]
For the runtime parts of the Effects API in MonoGame we need the metadata of the shaders like uniform and entry point names. We could parse it with HlslTools, but due to constraints on the target some names might change (e.g. old OpenGL versions don't separate textures and samplers), so we need the reflection data from SPIRV-Cross and DXC.

**Comment by JoaoBaptMG** [comment]
> The only syntax that's no longer supported is some of the SM1.0 syntax (e.g. the tex2d intrinsic). Other than that it's backwards compatible.

So I can compile SM4 and SM5 using DXC? Interesting; but still, I guess dxc itself cannot compile it, I might have to try to compile it myself.
Anyway, since SM4 and SM5 don't use the effect concept, it might be better to create another tool that would handle those separately.

On a side note, I could also try to add support for structured buffers/arbitrary constant buffers as well, since "raw parameters" aren't used in SM4/5.

**Comment by JoaoBaptMG** [comment]
> For the runtime parts of the Effects API in MonoGame we need the metadata of the shaders like uniform and entry point names.

The idea I was discussing on Discord was to `fxc` the shader files in WindowsDX and then use glslang/DXC plus SPIRV-Cross for the OpenGL/Vulkan/Metal/PS4 platforms, so that we can leverage Windows native tools where they are available (might be useful to, for example, use Visual Studio's own graphics debugger).

**Comment by Jjagg** [comment]
I don't follow, why'd you need fxc for that?

**Comment by JoaoBaptMG** [comment]
Because I haven't found anyway that fxc could be used to compile for shader models lower than 6.

**Comment by Jjagg** [comment]
> SPIRV-Cross can compile to older HLSL Shader Models.

So we can use the same pipeline for all targets.

**Comment by JoaoBaptMG** [comment]
Since we’re talking about modernising the shader tools, I guess I could drop this here too: https://github.com/mellinoe/ShaderGen

**Comment by Jjagg** [comment]
I wouldn't support ShaderGen officially, but if the pipeline works as intended it'd be possible to generate the HLSL shader with ShaderGen and feed that to our pipeline.

**Comment by tomspilman** [comment]
Closing this as it is part of #9188 now.  We will be moving away from MojoShader and to a singular modern cross-compiler tool for all shaders.
